core/templates/services/question-validation.service.ts

```diff
@@ -53,6 +53,34 @@ export class QuestionValidationService {
     );
   }
 
+  hasDuplicateRules(question: Question): boolean {
+    const answerGroups = question._stateData.interaction.answerGroups;
+    const ruleSet = new Set<string>();
+
+    for (const group of answerGroups) {
+      for (const rule of group.rules) {
+        const ruleHash = `${rule.type}:${this.normalizeRuleInputs(rule.inputs)}`;
+
+        if (ruleSet.has(ruleHash)) {
+          return true;
+        }
+        ruleSet.add(ruleHash);
+      }
+    }
+
+    return false;
+  }
+
+  normalizeRuleInputs(inputs: Record<string, unknown>): string {
+    const sortedKeys = Object.keys(inputs).sort();
+    const normalizedInputs: Record<string, unknown> = {};
```

Comment:
Are you sure this works in general, e.g. for interactions where the rule inputs are lists? I'm not sure if Records (which are meant to be immutable) can store those cases, and you don't have a test for that.

Please strictly-type this, don't just use unknown, otherwise you'll miss these use cases.

Reply:
Thank you for the feedback! You're right about the potential issue with handling lists in inputs. I will update the code to ensure it properly handles such cases. Additionally, I'll strictly type the normalizedInputs instead of using unknown to avoid any issues with unsupported use cases. I'll also add tests to cover scenarios with lists as inputs.